src/libasr/pass/replace_symbolic.cpp

```diff
@@ -0,0 +1,175 @@
+{
+public:
+    ReplaceSymbolicVisitor(Allocator &al_) :
+    PassVisitor(al_, nullptr) {
+        pass_result.reserve(al, 1);
+    }
+
+    bool symbolic_replaces_with_CPtr_Module = false;
+    bool symbolic_replaces_with_CPtr_Function = false;
+
+    void visit_Module(const ASR::Module_t &x) {
+        SymbolTable* current_scope_copy = current_scope;
+        current_scope = x.m_symtab;
+        for (auto &a : x.m_symtab->get_scope()) {
+            this->visit_symbol(*a.second);
+            if(symbolic_replaces_with_CPtr_Module){
+                std::string new_name = current_scope->get_unique_name("basic_new_stack");
+                if(current_scope->get_symbol(new_name)) return;
+                std::string header = "symengine/cwrapper.h";
+                SymbolTable *fn_symtab = al.make_new<SymbolTable>(current_scope);
+
+                Vec<ASR::expr_t*> args;
+                {
+                    args.reserve(al, 1);
+                    ASR::ttype_t *arg_type = ASRUtils::TYPE(ASR::make_CPtr_t(al, x.base.base.loc));
+                    ASR::symbol_t *arg = ASR::down_cast<ASR::symbol_t>(ASR::make_Variable_t(
+                        al, x.base.base.loc, fn_symtab, s2c(al, "x"), nullptr, 0, ASR::intentType::In,
+                        nullptr, nullptr, ASR::storage_typeType::Default, arg_type, nullptr,
+                        ASR::abiType::BindC, ASR::Public, ASR::presenceType::Required, true));
+                    fn_symtab->add_symbol(s2c(al, "x"), arg);
+                    args.push_back(al, ASRUtils::EXPR(ASR::make_Var_t(al, x.base.base.loc, arg)));
+                }
+                Vec<ASR::stmt_t*> body;
+                body.reserve(al, 1);
+
+                Vec<char *> dep;
+                dep.reserve(al, 1);
+
+                ASR::asr_t* new_subrout = ASRUtils::make_Function_t_util(al, x.base.base.loc,
+                    fn_symtab, s2c(al, new_name), dep.p, dep.n, args.p, args.n, body.p, body.n,
+                    nullptr, ASR::abiType::BindC, ASR::accessType::Public,
+                    ASR::deftypeType::Interface, nullptr, false, false, false,
+                    false, false, nullptr, 0, nullptr, 0, false, false, false, s2c(al, header));
+                ASR::symbol_t *new_symbol = ASR::down_cast<ASR::symbol_t>(new_subrout);
+                current_scope->add_symbol(new_name, new_symbol);
+                symbolic_replaces_with_CPtr_Module = false;
+            }
+        }
+        current_scope = current_scope_copy;
+    }
+
+    void visit_Function(const ASR::Function_t& x) {
+        ASR::Function_t &xx = const_cast<ASR::Function_t&>(x);
+        SymbolTable* current_scope_copy = current_scope;
+        current_scope = xx.m_symtab;
+        for (auto &item : current_scope->get_scope()) {
+            if (is_a<ASR::Variable_t>(*item.second)) {
+                this->visit_symbol(*item.second);
+                if(symbolic_replaces_with_CPtr_Function){
```

Comment:
Let's do this operation in the visit_variable itself.
And remove both visit_Module and visit_Function.
I think PassVisitor handles everything, Even `transform_stmts`.

Reply:
You can access the Function scope using `current_scope` and the Module scope using `current_scope->parent`.

Reply:
Store all the statements in pass_result.

Reply:
Ok so if we get rid of `visit_Module` and `visit_Function` ,
the scope for function is `x.m_parent_symtab` and that of module is `x.m_parent_symtab->parent` right ?

Reply:
PassVisitor would have assigned `current_scope` with `x.m_symtab` in `visit_Function` (in `asr.h`: `ASRPassBaseWalkVisitor`).
So you can use `current_scope` to access Function scope and `current_scope-> parent` to access Module scope.